### PR TITLE
Configurable onion request hops and wrong PoW/timestamp/swarm fixes for proxy/ORs

### PR DESCRIPTION
**Pull request by neuroscr**
- enables onion requests by default
- set onion request hops to 1
- fixes bug in proxy / onion requests not properly handling wrong PoW/timestamp/swarm
- log any sendMessage exceptions

### CODE REVIEW

File: js/modules/loki_message_api.js [hunk 1]
```diff
@@ -297,6 +305,7 @@ class LokiMessageAPI {
           // won't include parsing failures...
           successiveFailures = 0;
           if (messages.length) {
+            // log.debug(`loki_message:::_openRetrieveConnection - received ${messages.length} messages from ${nodeData.ip}:${nodeData.port}`)
```
Comment (Mikunj) [inline js/modules/loki_message_api.js:308]: Remove this log?

File: js/modules/loki_primitives.js [hunk 2]
```diff
@@ -109,7 +109,7 @@ function abortableIterator(array, iterator) {
     start: async serially => {
       let item = destructableList.pop();
       while (item && !abortIteration) {
-        // console.log('iterating on item', item);
+        // log.debug('iterating on item', item);
```
Comment (Mikunj) [inline js/modules/loki_primitives.js:112]: this too

File: js/modules/loki_rpc.js [hunk 3]
```diff
@@ -494,11 +521,41 @@ const lokiFetch = async (url, options = {}, targetNode = null) => {
         );
 
         if (result === BAD_PATH) {
-          log.error('[path] Error on the path');
+          const pathArr = [];
```
Comment (msgmaxim) [inline js/modules/loki_rpc.js:524]: How about we replace this with a funciton call that stringifies the path? (It is even used again below)

File: js/modules/loki_rpc.js [hunk 4]
```diff
@@ -494,11 +521,41 @@ const lokiFetch = async (url, options = {}, targetNode = null) => {
...
+            `[path] Error on the path: ${pathArr.join(', ')} to ${
+              targetNode.ip
+            }:${targetNode.port}`
+          );
           lokiSnodeAPI.markPathAsBad(path);
+          return false;
+        }
+
+        // result maybe false
+        if (result) {
```
Comment (Mikunj) [inline js/modules/loki_rpc.js:538]: We can use [`result.ok`](https://developer.mozilla.org/en-US/docs/Web/API/Response/ok) to determine if the status code is within the success range (200 - 299)

File: js/modules/loki_rpc.js [hunk 5]
```diff
@@ -527,8 +584,17 @@ const lokiFetch = async (url, options = {}, targetNode = null) => {
         // pass the false value up
         return false;
       }
+
+      // result maybe false
+      if (result) {
```
Comment (msgmaxim) [inline js/modules/loki_rpc.js:589]: `if` -> `else if `?

File: preload.js [hunk 6]
```diff
@@ -414,7 +414,8 @@ window.lokiFeatureFlags = {
   privateGroupChats: true,
   useSnodeProxy: !process.env.USE_STUBBED_NETWORK,
```
Comment (Mikunj) [inline preload.js:415]: Might be wise to set this to `false`? 
Once we have onion requests we can strip out all snode proxying code

File: preload.js [hunk 7]
```diff
@@ -440,14 +441,7 @@ if (
...
   window.lokiFeatureFlags = {};
-  window.lokiSnodeAPI = {
-    refreshSwarmNodesForPubKey: () => [],
-    getFreshSwarmNodes: () => [],
-    updateSwarmNodes: () => {},
-    updateLastHash: () => {},
-    getSwarmNodesForPubKey: () => [],
-    buildNewOnionPaths: () => [],
-  };
+  window.lokiSnodeAPI = {}; // no need stub out each function here
```
Comment (Mikunj) [inline preload.js:444]: Do integration tests still work this this change?

File: preload.js (hunk 6 above)
Comment (neuroscr) [reply preload.js:415]: file server code can still use proxy where it can't use OR

File: js/modules/loki_snode_api.js [hunk 8]
```diff
@@ -348,19 +348,28 @@ class LokiSnodeAPI {
     const guards = _.shuffle(this.guardNodes);
 
     // Create path for every guard node:
-
-    // Each path needs 2 nodes in addition to the guard node:
-    const maxPath = Math.floor(Math.min(guards.length, otherNodes.length / 2));
+    const needPaths = window.lokiFeatureFlags.onionRequestPaths - 1;
```
Comment (msgmaxim) [inline js/modules/loki_snode_api.js:351]: `needPaths`? How about `needNodesPerPath`? Also `onionRequestPaths` below should be called something like `nodesPerPath`.

File: preload.js (hunk 6 above)
Comment (msgmaxim) [reply preload.js:415]: We should remove proxy code when we no longer use it, but I think the file server is not on onion requests yet.

File: preload.js (hunk 7 above)
Comment (neuroscr) [reply preload.js:444]: yes
```
RMacPro:lm-n-fileproxy user2$ yarn test-integration
yarn run v1.17.3
$ ELECTRON_DISABLE_SANDBOX=1 mocha --exit --timeout 10000 integration_test/integration_test.js


  Window Test and Login
    ✓ registration: opens one window
    ✓ registration: window title is correct
    ✓ registration: can restore from seed (11046ms)
    ✓ registration: can create new account (7735ms)
    ✓ registration: can delete account when logged in (16418ms)

  Open groups
    ✓ openGroup: works with valid open group url
    ✓ openGroup: cannot join two times the same open group
    ✓ openGroup: can send message to open group (15590ms)

  Add friends
    ✓ addFriends: can add a friend by sessionID (8189ms)

  Link Device
    ✓ linkDevice: link two desktop devices
    ✓ linkDevice: unlink two devices (14640ms)

  Closed groups
    ✓ closedGroup: can create a closed group with a friend and send/receive a message (9899ms)
```

File: js/modules/loki_rpc.js [hunk 9]
```diff
@@ -493,12 +520,41 @@ const lokiFetch = async (url, options = {}, targetNode = null) => {
           fetchOptions.body
         );
 
+        const getPathString = pathObjArr => {
+          const pathArr = [];
+          pathObjArr.forEach(node => {
+            pathArr.push(`${node.ip}:${node.port}`);
+          });
+          return pathArr.join(', ');
```
Comment (msgmaxim) [inline js/modules/loki_rpc.js:528]: A nitpick, but could be done with map and join.

File: js/modules/loki_rpc.js (hunk 4 above)
Comment (neuroscr) [reply js/modules/loki_rpc.js]: the proxy/OR don't include the `ok` property when we fake a response object

File: js/modules/loki_rpc.js [hunk 10]
```diff
@@ -477,14 +512,6 @@ const lokiFetch = async (url, options = {}, targetNode = null) => {
         const thisIdx = onionReqIdx;
         onionReqIdx += 1;
 
-        log.debug(
```
Comment (msgmaxim) [inline js/modules/loki_rpc.js:480]: Is this just gone now? It is on `debug` so I think we could keep it.

File: js/modules/loki_snode_api.js (hunk 8 above)
Comment (neuroscr) [reply js/modules/loki_snode_api.js:351]: Going with `onionRequestHops` and `nodesNeededPerPaths`

File: js/modules/loki_rpc.js (hunk 10 above)
Comment (neuroscr) [reply js/modules/loki_rpc.js:480]: yes, it's just creates a lot of noise. It's not an indication of a problem. We send a lot of requests over ORs. And now we just include this path on any error/exception

File: js/modules/loki_rpc.js (hunk 4 above)
Comment (Mikunj) [reply js/modules/loki_rpc.js]: Ah ok fair enough 👍

File: preload.js (hunk 6 above)
Comment (Mikunj) [reply preload.js:415]: 👍

**Pull request merged**